### ISSUE TITLE
Have `Enumerable.max` with optional argument `n` to return `n` maximum elements.

### ISSUE THREAD

**Issue by imsofi** [crystal-lang/crystal#12817]
## Feature Request

Allow `Enumerable.max` to take an optional argument `n`, which will return the first `n` maximum elements in a collection.

This is inspired by similar functionality for the Enumerable module in Ruby[^1].

#### Example:
```crystal
(1..10).max(3)  # => [10, 9, 8]
```

#### Alternatives:
One can use the pattern of `sort.reverse.first(3)` to get the same result.

#### Reasoning:
Using `.sort.reverse.first(m)` will run in roughly `O(n log n)` time, due to the use of a full array sort. While the function `.max(m)` will realistically run in `O(n)` time, as long as `m` is set reasonably low.

#### Backwards compatibility:
This should not break backwards compatibility, as there currently `Enumerable.max` does not have any optional arguments.

[^1]: https://docs.ruby-lang.org/en/3.1/Enumerable.html#method-i-max

**Comment by asterite** [comment]
Oh, nice! I needed this for Advent of Code. I was wondering if that API would make sense, and I thought it didn't exist in Ruby, but apparently it does! So 👍 from me.

**Comment by asterite** [comment]
By the way, if you do this we should also do it for `min`.

**Comment by imsofi** [comment]
I may or may not have opened this issue here due to wanting a simpler method to find a set of maximum's for Advent of Code myself. :sweat_smile:

**Comment by yxhuvud** [comment]
Potentially also apply to `m??_of` and `m??_by`. 

You won't be the first to have opened issues in response to AoC, I have done that too a couple of times. :)

I'm almost certain this is a duplicate, by the way. But I can't find it. @straight-shoota I'm almost certain you also commented on it. Do you have any idea what I'm referring to?

For implementation, I'd suggest some sort of heap (where things are taken off the heap when needed), though a sorted array might be good enough, especially for small n.

That said a heap would be nice for other AoC tasks :smile:

**Comment by asterite** [comment]
@yxhuvud Yeah, after thinking about it for a while I realized doing this is far from trivial. Or, put another way, doing this efficiently is not that simple. It seems there's a Quick Search algorithm that's similar to Quick Sort, I think Ruby might used that for the Array case. For the Enumerable case I'm not sure, probably a heap.

**Comment by straight-shoota** [comment]
@yxhuvud I don't recall any prior discussion, sorry.

**Comment by imsofi** [comment]
> I'm almost certain this is a duplicate

@yxhuvud I also tried looking for a duplicate before i opened this issue, but could not find anything similar.

**Comment by HertzDevil** [comment]
Ruby MRI uses [quickselect](https://github.com/ruby/ruby/blob/cada537040743cbe49aac6740816d648ca0d3fb6/enum.c#L1732-L1813) followed by [quicksort](https://github.com/ruby/ruby/blob/cada537040743cbe49aac6740816d648ca0d3fb6/enum.c#L1884-L1902). The case for general `Enumerable`s simply materializes all elements.

**Comment by straight-shoota** [comment]
The non-raising variants `#max?(k)` and `#min?(k)` haven't been discussed here yet.
I suppose it makes sense to have them, but the exact semantics need to be clarified.

The existing methods without parameter return `nil` only when the collection is empty. I think that would make sense for the new ones as well.
It should still raise if the input parameter is invalid or if elements are not comparable.

**Comment by asterite** [comment]
`max?(k)` and `min?(k)` don't make sense. `max(k)` means "give me the maximum k elements, and if you have less than k elements, well, give me what you have". That means that if the enumerable is empty, empty is returned. This is the same as in Ruby.

There's no `first?(n)` method in Array for that same reason.
